File: test/scripts/02-busy-wait/incorrect.js
#!/usr/bin/env node
/**
 * INCORRECT: Busy-wait loop - burns 100% CPU
 */
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const sab = new SharedArrayBuffer(4);
const stopFlag = new Int32Array(sab);

function busyWait(deadline) {
  while (Date.now() < deadline && !Atomics.load(stopFlag, 0)) {}
}

console.log('INCORRECT: Busy-wait');
const deadline = Date.now() + DURATION_SEC * 1000;
const t = setInterval(() => { Atomics.store(stopFlag, 0, 1); clearInterval(t); console.log('Stopped'); }, DURATION_SEC * 1000);
process.on('SIGINT', () => Atomics.store(stopFlag, 0, 1));
busyWait(deadline);

File: test/scripts/02-busy-wait/incorrect.py
#!/usr/bin/env python3
"""
INCORRECT: Busy-wait (spin loop) - burns 100% CPU, never yields.
Resume: "Avoid Busy-Wait Loops"
Run: DURATION_SEC=60 python incorrect.py
"""
import os
import signal
import threading
import time

NUM_WAITERS = 4
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def busy_wait_until(deadline):
    while time.perf_counter() < deadline and not stop:
        pass  # Busy-wait: 100% CPU

def main():
    global stop
    print("INCORRECT: Busy-wait. Run: top -p $(pgrep -f busy)")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    deadline = time.perf_counter() + DURATION_SEC
    threads = [threading.Thread(target=busy_wait_until, args=(deadline,)) for _ in range(NUM_WAITERS)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    print("Stopped")

if __name__ == "__main__":
    main()
